metals/src/main/scala/scala/meta/internal/bsp/BspConnector.scala

```diff
@@ -0,0 +1,200 @@
+import scala.meta.internal.builds.SbtBuildTool
+import scala.meta.internal.builds.SbtServer
+import scala.meta.internal.metals.BloopServers
+import scala.meta.internal.metals.BuildServerConnection
+import scala.meta.internal.metals.Messages
+import scala.meta.internal.metals.Messages.BspSwitch
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.Tables
+import scala.meta.internal.metals.UserConfiguration
+import scala.meta.io.AbsolutePath
+
+import ch.epfl.scala.bsp4j.BspConnectionDetails
+import com.google.common.collect.ImmutableList
+import org.eclipse.lsp4j.services.LanguageClient
+
+object BspConnector {
+  final val BLOOP_SELECTED = "BLOOP"
+}
+
+class BspConnector(
+    bloopServers: BloopServers,
+    sbtServer: SbtServer,
+    bspServers: BspServers,
+    buildTools: BuildTools,
+    client: LanguageClient,
+    tables: Tables,
+    userConfig: () => UserConfiguration
+)(implicit ec: ExecutionContext) {
+
+  def resolve(): BspResolvedResult = {
+    resolveExplicit().getOrElse {
+      if (buildTools.isBloop) ResolvedBloop
+      else bspServers.resolve()
+    }
+  }
+
+  private def resolveExplicit(): Option[BspResolvedResult] = {
+    tables.buildServers.selectedServer().flatMap { sel =>
+      if (sel == BspConnector.BLOOP_SELECTED) Some(ResolvedBloop)
+      else
+        bspServers
+          .findAvailableServers()
+          .find(_.getName == sel)
+          .map(x => ResolvedBspOne(x))
+    }
+  }
+
+  def connect(
+      workspace: AbsolutePath,
+      userConfiguration: UserConfiguration
+  )(implicit ec: ExecutionContext): Future[Option[BspSession]] = {
+    def connect(
+        workspace: AbsolutePath
+    ): Future[Option[BuildServerConnection]] = {
+      scribe.info("Attempting to connect to the build server...")
+      resolve() match {
+        case ResolvedNone =>
+          scribe.info("No build server found")
+          Future.successful(None)
+        case ResolvedBloop =>
+          scribe.info(
+            "Attempting to start a bloop connection from previous choice"
+          )
+          bloopServers.newServer(workspace, userConfiguration).map(Some(_))
+        case ResolvedBspOne(details) if details.getName == SbtBuildTool.name =>
+          scribe.info(
+            s"Attempting to start a new connection to ${details.getName()} from previous choice..."
+          )
+          // NOTE: we explicity start another sbt process here as simply using newServer
+          // here wil indeed start sbt, but not correctly star the bsp sessions. So before
+          // we start the session we ensure that sbt is running, and then connect.
+          val (_, processHandler) = sbtServer.runSbtShell()
+          processHandler.initialized.future.flatMap { _ =>
+            scribe.info(
+              s"sbt up and running, attempting to start a bsp session..."
+            )
+            bspServers.newServer(workspace, details).map(Some(_))
+          }
```

Comment:
Depending on what happens here in https://github.com/sbt/sbt/issues/5993, this may not actually be needed. The issue I was hitting on and reported here https://github.com/scalameta/metals/issues/2150 may have been a red herring not due to file encoding but rather not being able to correctly parse the output when attempting to make a bsp connect. If that's the case, we don't actually need to start a new sbt shell here before making the bsp connect, and instead just allow that process handler to handle the whole thing.

Reply:
So this actually works as expected in 1.4.1 without needed the `runSbtShell` call here, which is great, but it will be a problem for anyone trying with 1.4.0. Not sure to leave it in here or remove it.